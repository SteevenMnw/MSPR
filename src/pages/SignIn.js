import React from "react";
import {
  View,
  StyleSheet,
  Text,
  KeyboardAvoidingView,
  TouchableWithoutFeedback,
  Keyboard,
  TouchableOpacity,
} from "react-native";
import { Input, Button } from "react-native-elements";
import Icon from "react-native-vector-icons/Ionicons";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { getUserByEmailAndPassword } from "../API/API_Access";
import * as Crypto from "expo-crypto";

class SignIn extends React.Component {

  //initialize variables
  constructor(props) {
    super(props);
    this.state = { dataSource: [], email: "", passwd: "", error: "" };
  }

  //Update variables when changed
  updateEmail = (email) => {
    this.setState({ email: email });
  };

  updatePasswd = (passwd) => {
    this.setState({ passwd: passwd });
  };

  updateError = (error) => {
    this.setState({ error: error });
  };

  //Set the User data from API to Session to use it in another page
  setUserSession = async (value, pswd) => {
    try {
      AsyncStorage.clear();
      const jsonValue = JSON.stringify(value);
      await AsyncStorage.setItem("user", jsonValue);
      const jsonValuePswd = JSON.stringify(pswd);
      await AsyncStorage.setItem("pswd", jsonValuePswd);
    } catch (e) {
      console.log(e);
    }
  };

  //Function to connect the User
  getUser = () => {
    try {
      //Checking the accuracy of the information
      email = this.state.email;
      passwd = this.state.passwd;
      if (email && passwd) {
        //Crypt password
        const value = Crypto.digestStringAsync(
          Crypto.CryptoDigestAlgorithm.SHA256,
          passwd
        );
        value.then((newPasswd) => {
          //Call the API with parameters to check if the information are correct
          getUserByEmailAndPassword(email, newPasswd)
            .then((data) => {
              this.setState({ dataSource: data });
              if (this.state.dataSource) {
                //Call the UserSession function
                this.setUserSession(this.state.dataSource, this.state.passwd);
                //Go to the Home page
                this.props.navigation.navigate("Home");
              }
            })
            .catch(() =>
              this.updateError("Votre mail ou mot de passe est invalide")
            );
        });
      } else if (passwd == "" && email) {
        this.updateError("Veuillez insérer votre mot de passe.");
      } else if (email == "" && passwd) {
        this.updateError("Veuillez insérer votre mail.");
      } else {
        this.updateError("Veuillez renseigner toutes les informations.");
      }
    } catch (e) {
      console.log(e);
    }
  };

  render() {
    return (
      //Prevents the keyboard display from shifting everything up 
      <KeyboardAvoidingView
        testID="test_keyboardAvoidingView"
        style={styles.container}
      >
        {/* Allows you to exit keyboard mode by clicking elsewhere */}
        <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
          <View testID="test_view_mainContainer" style={styles.main_container}>
            <View testID="test_view_input" style={styles.Input}>
              <Input
                testID="test_input_email"
                placeholder="Mail"
                leftIcon={<Icon name="mail-outline" size={24} color="black" />}
                onChangeText={this.updateEmail}
                value={this.state.email}
              />
              <Input
                testID="test_input_password"
                secureTextEntry={true}
                placeholder="Mot de passe"
                leftIcon={
                  <Icon name="lock-closed-outline" size={24} color="black" />
                }
                onChangeText={this.updatePasswd}
                value={this.state.passwd}
              />
            </View>
            <View testID="test_view_error-button">
              <Text testID="test_text_error" style={styles.Error}>
                {this.state.error}
              </Text>
              <Button
                testID="test_button_signIn"
                buttonStyle={{
                  width: 150,
                  alignSelf: "center",
                  backgroundColor: "#ba473c",
                }}
                title="Se connecter"
                onPress={this.getUser}
              />
            </View>
            <View
              testID="test_view_navigation"
              style={{ alignSelf: "center", paddingTop: 30 }}
            >
              <Text testID="test_text_beforeNavigation">
                Vous n'avez pas encore de compte ?
              </Text>
              <TouchableOpacity
                testID="test_touchableOpacity_navigation"
                style={{ alignSelf: "center" }}
                //Go to the SignUp page
                onPress={() => this.props.navigation.navigate("SignUp")}
              >
                <Text
                  testID="test_text_navigation"
                  style={{ color: "#ba473c" }}
                >
                  S'inscrire
                </Text>
              </TouchableOpacity>
            </View>
          </View>
        </TouchableWithoutFeedback>
      </KeyboardAvoidingView>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  main_container: {
    flex: 1,
    justifyContent: "center",
  },
  Input: {
    paddingLeft: 70,
    paddingRight: 70,
  },
  Error: {
    paddingBottom: 25,
    alignSelf: "center",
    color: "red",
    fontSize: 17,
  },
});

export default SignIn;
