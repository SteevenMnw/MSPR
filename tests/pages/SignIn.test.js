import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import SignIn from "../../src/pages/SignIn";
import {fireEvent, render} from '@testing-library/react-native'
import {expect, it} from '@jest/globals';

//global.fetch = jest.fn(() => new Promise(resolve => resolve()));
//jest.mock('react-native-gesture-handler', () => {});

it('renders correctly', () => {
  const tree = renderer.create(<SignIn />).toJSON();
  expect(tree).toMatchSnapshot();
});

//jest.mock('@react-native-community/async-storage', () => ({setItem: jest.fn()}))

it("renders default elements and styles",() => {
  const {getAllByText, getByPlaceholderText, getByText, getByTestId} = render(<SignIn />)

  getByPlaceholderText("Mail");
  getByPlaceholderText("Mot de passe");

  expect(getAllByText("Se connecter").length).toBe(1);
  expect(getAllByText("Vous n'avez pas encore de compte ?").length).toBe(1);
  expect(getAllByText("S'inscrire").length).toBe(1);


  expect(getByTestId(/test_keyboardAvoidingView/i).props.style).toMatchObject({
    "flex": 1,
  }, "test KeyboardAvoidingView style")
  expect(getByTestId(/test_view_mainContainer/i).props.style).toMatchObject({
    "flex": 1,
    "justifyContent": "center",
  }, "test MainContainer style")
  expect(getByTestId(/test_view_input/i).props.style).toMatchObject({
    "paddingLeft": 70,
    "paddingRight": 70,
  }, "test Input style")
  expect(getByPlaceholderText(/Mail/i).parent.props.style).toMatchObject({
    "alignItems": "center",
    "borderBottomWidth": 1,
    "borderColor": "#86939e",
    "flexDirection": "row",
  }, "test PlaceHolder Input Mail style")
  expect(getByTestId(/test_input_mail/i).props.style).toMatchObject({
    "alignSelf": "center",
    "color": "#242424",
    "flex": 1,
    "fontSize": 18,
    "minHeight": 40,
  }, "test InputMail style")
  expect(getByPlaceholderText(/Mot de passe/i).parent.props.style).toMatchObject({
    "alignItems": "center",
    "borderBottomWidth": 1,
    "borderColor": "#86939e",
    "flexDirection": "row",
  }, "test PlaceHolder Input MotDePasse style")
  expect(getByTestId(/test_input_motDePasse/i).props.style).toMatchObject({
    "alignSelf": "center",
    "color": "#242424",
    "flex": 1,
    "fontSize": 18,
    "minHeight": 40,
  }, "test InputMotDePasse style")
  expect(getByTestId(/test_view_error-button/i).props.style).toBeUndefined()
  expect(getByTestId(/test_text_error/i).props.style).toMatchObject({
    "alignSelf": "center",
    "color": "red",
    "fontSize": 17,
    "paddingBottom": 25,
  }, "test Text Error style")
  expect(getByTestId(/test_button_signIn/i).props.style).toMatchObject({
    "opacity": 1,
  }, "test View Button SignIn style")
  expect(getByText(/Se connecter/i).parent.props.style).toMatchObject({
    "alignItems": "center",
    "alignSelf": "center",
    "backgroundColor": "#ba473c",
    "borderColor": "#2089dc",
    "borderRadius": 3,
    "borderWidth": 0,
    "flexDirection": "row",
    "justifyContent": "center",
    "padding": 8,
    "width": 150,
  }, "test Button SignIn style")
  expect(getByText(/Se connecter/i).props.style).toMatchObject({
    "color": "white",
    "fontSize": 18,
    "paddingVertical": 1,
    "textAlign": "center",
  }, "test Text Button SignIn style")
  expect(getByTestId(/test_view_navigation/i).props.style).toMatchObject({
    "alignSelf": "center",
    "paddingTop": 30,
  }, "test View Navigation style")
  expect(getByTestId(/test_text_beforeNavigation/i).props.style).toBeUndefined()
  expect(getByText(/Vous n'avez pas encore de compte ?/i).props.style).toBeUndefined()
  expect(getByTestId(/test_touchableOpacity_navigation/i).props.style).toMatchObject({
    "alignSelf": "center",
    "opacity": 1,
  }, "test touchableOpacity Navigation style")
  expect(getByTestId(/test_text_navigation/i).props.style).toMatchObject({
    "color": "#ba473c",
  }, "test Text Navigation style1")
  expect(getByText(/S'inscrire/i).props.style).toMatchObject({
    "color": "#ba473c",
  }, "test Text Navigation style2")
});

/*it("shows invalid inputs messages",() => {
  const {getByText} = render(<SignIn/>)
  
});*/
