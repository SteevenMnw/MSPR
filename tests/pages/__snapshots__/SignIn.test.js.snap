// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
  testID="test_keyboardAvoidingView"
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
    testID="test_view_mainContainer"
  >
    <View
      style={
        Object {
          "paddingLeft": 70,
          "paddingRight": 70,
        }
      }
      testID="test_view_input"
    >
      <View
        style={
          Object {
            "paddingHorizontal": 10,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 1,
              "borderColor": "#86939e",
              "flexDirection": "row",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "marginVertical": 4,
                "paddingRight": 4,
              }
            }
          >
            <Text />
          </View>
          <TextInput
            allowFontScaling={true}
            editable={true}
            onChangeText={[Function]}
            placeholder="Mail"
            placeholderTextColor="#86939e"
            rejectResponderTermination={true}
            style={
              Object {
                "alignSelf": "center",
                "color": "#242424",
                "flex": 1,
                "fontSize": 18,
                "minHeight": 40,
              }
            }
            testID="test_input_mail"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <Text
          style={
            Object {
              "color": "#ff190c",
              "fontSize": 12,
              "margin": 5,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "paddingHorizontal": 10,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 1,
              "borderColor": "#86939e",
              "flexDirection": "row",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "marginVertical": 4,
                "paddingRight": 4,
              }
            }
          >
            <Text />
          </View>
          <TextInput
            allowFontScaling={true}
            editable={true}
            onChangeText={[Function]}
            placeholder="Mot de passe"
            placeholderTextColor="#86939e"
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Object {
                "alignSelf": "center",
                "color": "#242424",
                "flex": 1,
                "fontSize": 18,
                "minHeight": 40,
              }
            }
            testID="test_input_motDePasse"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <Text
          style={
            Object {
              "color": "#ff190c",
              "fontSize": 12,
              "margin": 5,
            }
          }
        />
      </View>
    </View>
    <View
      testID="test_view_error-button"
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "red",
            "fontSize": 17,
            "paddingBottom": 25,
          }
        }
        testID="test_text_error"
      >
        
      </Text>
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            undefined,
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="test_button_signIn"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#ba473c",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
                "width": 150,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "white",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Se connecter
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignSelf": "center",
          "paddingTop": 30,
        }
      }
      testID="test_view_navigation"
    >
      <Text
        testID="test_text_beforeNavigation"
      >
        Vous n'avez pas encore de compte ?
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "center",
            "opacity": 1,
          }
        }
        testID="test_touchableOpacity_navigation"
      >
        <Text
          style={
            Object {
              "color": "#ba473c",
            }
          }
          testID="test_text_navigation"
        >
          S'inscrire
        </Text>
      </View>
    </View>
  </View>
</View>
`;
